### DISCUSSION TITLE
Custom Table Name

### DISCUSSION THREAD

**Discussion by Tejesh-Duvvuru** [hasura/graphql-engine#9419] [Question]
I am using cloud account my problem is with table names when we created table name we use - (table1-dev) like that but now problem is hasura( graphql) not recognizing them. I can't rename table is production because it will cause so many problems. Is there any use to custom table name hasura.

**Comment by meetzaveri** [comment]
Hello @Tejesh-Duvvuru 

Please follow these steps in order to track your table with custom name.

#### Step 1: First, untrack the table 
If your table is already tracked, then first you’ll need to untrack that table. You can use `pg_untrack_table` metadata API to untrack the table. API Reference - https://hasura.io/docs/latest/api-reference/metadata-api/table-view/#metadata-pg-untrack-table

Example payload
```json
{
    "type": "pg_untrack_table",
    "args": {
        "table": {
            "schema": "public",
            "name": “Devices”
         },
        "source": "default",
        "cascade": true
    }
}
```



#### Step 2: Track the table with new custom name
Then to make it track with a new custom name, use `pg_track_table` metadata API . In that, pass custom name to `custom_name` field under configuration object in metadata API payload. See the below example. API reference - https://hasura.io/docs/latest/api-reference/metadata-api/table-view/#metadata-pg-track-table

In the above API ref docs, see the 2nd paragraph where it mentions the use case of custom name.

Example payload
```json
{
   "type": "pg_track_table",
   "args": {
     "source": "default",
     "table": "Devices",
     "configuration": {
        "custom_name": "my_devices"
     }
   }
}
```

#### Making API requests to `v1/metadata` endpoint
Suppose your graphql endpoint is `https://new-project.hasura.app/v1/graphql` then, you need to make `POST` API requests to `v1/metadata` endpoints such as `https://new-project.hasura.app/v1/metadata` . You can perform this in POSTMAN or any other API tools. Make sure to pass the admin secret or necessary headers to interact with the hasura server. In my case, I used `x-hasura-admin-secret` in header for just this purpose.
 
#### Note - Skip the 1st step if you haven't tracked your database tables yet.

**Reply by Tejesh-Duvvuru** [reply]
Thanks  for responding @meetzaveri , but while doing the above process it given like this.
![image](https://user-images.githubusercontent.com/101654993/217230214-7bcba1e4-f3df-4f11-a536-0aa0c700faeb.png)


{
    "code": "validation-failed",
    "error": "cannot include \"userProfileTable-devs_pkey\" in the GraphQL schema because  it is not a valid GraphQL identifier",
    "path": "$.args"
}

**Reply by meetzaveri** [reply]
@Tejesh-Duvvuru 
Looking at the error response, it seems you haven't untracked the table `userProfileTable-devs` because it shows in the error itself as `userProfileTable-devs_pkey`. Can you please untrack it first, as suggested in step 1 ?

**Comment by Tejesh-Duvvuru** [comment]
@meetzaveri 
The table is in untracked only
![image](https://user-images.githubusercontent.com/101654993/217235220-dd9a4e6b-dce5-4ba8-aa06-30cafc204047.png)

{
    "code": "already-untracked",
    "error": "view/table already untracked: \"userProfileTable-devs\"",
    "path": "$.args"
}

**Reply by meetzaveri** [reply]
Oh right. I see this table contains `-` hyphen in its name. If you try to create a table which includes hyphen `-`, it will throw out an error saying `Table name cannot contain special characters. It can have letters, numbers and _ (cannot start with numbers)`

**Reply by Tejesh-Duvvuru** [reply]
@meetzaveri 

Thank you very much for the information.

**Reply by meetzaveri** [reply]
I don't know how on the first place that table got created with hyphen symbol(`-`) in table name. Can you try to rename it and then check if it works ?

**Reply by Tejesh-Duvvuru** [reply]
@meetzaveri 

After renaming table name "userProfileTabledevs" also showing like this error
Adding existing table/view failed
cannot include "userProfileTable-devs_pkey" in the GraphQL schema because it is not a valid GraphQL identifier

![image](https://user-images.githubusercontent.com/101654993/217255844-0728b55d-9acc-4123-9aac-dc29facaeca6.png)

**Reply by meetzaveri** [reply]
It seems like DB is still not updated with the new table name, though. Try checking list of table with `psql` command by running `\dt` which lists all the table under current schema.

**Reply by Tejesh-Duvvuru** [reply]
@meetzaveri 

I got it. In db the primary key is in the this(userProfileTable-devs_pkey) format  so it throwing error, now resolved it.

Thank you very much.
